src/app/clusters/mode-base-server/mode-base-server.cpp

```diff
@@ -0,0 +1,487 @@
+}
+
+Status Instance::UpdateCurrentMode(uint8_t aNewMode)
+{
+    if (!IsSupportedMode(aNewMode))
+    {
+        return Protocols::InteractionModel::Status::InvalidValue;
+    }
+    uint8_t oldMode = mCurrentMode;
+    mCurrentMode    = aNewMode;
+    if (mCurrentMode != oldMode)
+    {
+        // Write new value to persistent storage.
+        GetAttributePersistenceProvider()->WriteScalarValue(
+            ConcreteAttributePath(mEndpointId, mClusterId, Attributes::CurrentMode::Id), mCurrentMode);
+        // The Administrator Commissioning cluster is always on the root endpoint.
+        MatterReportingAttributeChangeCallback(mEndpointId, mClusterId, Attributes::CurrentMode::Id);
+    }
+    return Protocols::InteractionModel::Status::Success;
+}
+
+Status Instance::UpdateStartUpMode(DataModel::Nullable<uint8_t> aNewStartUpMode)
+{
+    if (!aNewStartUpMode.IsNull())
+    {
+        if (!IsSupportedMode(aNewStartUpMode.Value()))
+        {
+            return Protocols::InteractionModel::Status::InvalidValue;
+        }
+    }
+    DataModel::Nullable<uint8_t> oldStartUpMode = mStartUpMode;
+    mStartUpMode                                = aNewStartUpMode;
+    if (mStartUpMode != oldStartUpMode)
+    {
+        // Write new value to persistent storage.
+        GetAttributePersistenceProvider()->WriteScalarValue(
+            ConcreteAttributePath(mEndpointId, mClusterId, Attributes::StartUpMode::Id), mStartUpMode);
+        // The Administrator Commissioning cluster is always on the root endpoint.
+        MatterReportingAttributeChangeCallback(mEndpointId, mClusterId, Attributes::StartUpMode::Id);
+    }
+    return Protocols::InteractionModel::Status::Success;
+}
+
+Status Instance::UpdateOnMode(DataModel::Nullable<uint8_t> aNewOnMode)
+{
+    if (!aNewOnMode.IsNull())
+    {
+        if (!IsSupportedMode(aNewOnMode.Value()))
+        {
+            return Protocols::InteractionModel::Status::InvalidValue;
+        }
+    }
+    DataModel::Nullable<uint8_t> oldOnMode = mOnMode;
+    mOnMode                                = aNewOnMode;
+    if (mOnMode != oldOnMode)
+    {
+        // Write new value to persistent storage.
+        GetAttributePersistenceProvider()->WriteScalarValue(ConcreteAttributePath(mEndpointId, mClusterId, Attributes::OnMode::Id),
+                                                            mOnMode);
+        // The Administrator Commissioning cluster is always on the root endpoint.
+        MatterReportingAttributeChangeCallback(mEndpointId, mClusterId, Attributes::OnMode::Id);
+    }
+    return Protocols::InteractionModel::Status::Success;
+}
+
+DataModel::Nullable<uint8_t> Instance::GetStartUpMode() const
+{
+    return mStartUpMode;
+}
+
+DataModel::Nullable<uint8_t> Instance::GetOnMode() const
+{
+    return mOnMode;
+}
+
+uint8_t Instance::GetCurrentMode() const
+{
+    return mCurrentMode;
+}
+
+bool Instance::IsSupportedMode(uint8_t modeValue)
+{
+    for (uint8_t i = 0; i < NumberOfModes(); i++)
```

Comment:
If we ditch NumberOfModes(), this would just loop until GetModeValueByIndex returns CHIP_ERROR_NOT_FOUND.

Reply:
Modified [here](https://github.com/project-chip/connectedhomeip/pull/27504/commits/20a9f05d48d9f382613b0a5054aa20346268b1d2).